File: kotlin/src/main/kotlin/aether/voice/GroupVoiceCallService.kt
// SPDX-License-Identifier: MIT

package aether.voice

import aether.AetherConstants
import aether.protocol.MeshPacket
import aether.protocol.PacketType
import aether.routing.MeshSender
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

// ─────────────────────────────────────────────────────────────────────────────
// Domain types
// ─────────────────────────────────────────────────────────────────────────────

enum class GroupVoiceCallState { Invited, Active, Ended }

data class GroupVoiceCallSession(
    val id: UUID,
    val hostUhid: String,
    val members: MutableSet<String> = ConcurrentHashMap.newKeySet(),
    var state: GroupVoiceCallState = GroupVoiceCallState.Invited,
    /** Current key-generation counter. Host increments on each join/leave. */
    @Volatile var keyGeneration: Int = 0
) {
    val keyGenerationUInt: UInt get() = keyGeneration.toUInt()
}

/**
 * Group voice call service (up to [AetherConstants.MAX_GROUP_VOICE_MEMBERS] members).
 *
 * Wire format — GroupVoiceFrame binary payload:
 *   [16] CallId      (UUID RFC4122 big-endian)
 *   [4]  Sequence    (UInt32 little-endian)
 *   [8]  TimestampMs (Int64 little-endian)
 *   [1]  IsSilence   (0 or 1)
 *   [4]  KeyGeneration (UInt32 little-endian)
 *   [N]  EncodedPayload
 *
 * Signaling uses GroupVoiceSignalingMessage JSON (snake_case).
 *
 * Key rotation: when a member joins or leaves the host increments keyGeneration
 * and broadcasts a "key_rotation" signaling packet so all members re-key.
 *
 * Priority: 64 for frames, 32 for signaling.
 */
class GroupVoiceCallService(private val sender: MeshSender) {

    private val sessions = ConcurrentHashMap<UUID, GroupVoiceCallSession>()
    @Volatile private var frameSequence: Int = 0

    /** Fired when this node is invited to a group call. */
    var onInvited: ((GroupVoiceCallSession) -> Unit)? = null

    /** Fired when a member joins, leaves, is kicked, or the call ends. */
    var onMembershipChanged: ((GroupVoiceCallSession) -> Unit)? = null

    /** Fired when a new key generation is announced (re-keying needed). */
    var onKeyRotation: ((UUID, UInt) -> Unit)? = null

    /** Fired when a group voice frame arrives. */
    var onFrameReceived: ((UUID, String, ByteArray, Boolean, UInt) -> Unit)? = null

    // ─────────────────────────────────────────────────────────────────────
    // Host API
    // ─────────────────────────────────────────────────────────────────────

    /**
     * (Host) Invite [memberUhids] into an existing or new call [callId].
     */
    suspend fun invite(callId: UUID, memberUhids: List<String>) {
        require(memberUhids.isNotEmpty()) { "memberUhids must not be empty" }
        val session = sessions.getOrPut(callId) {
            GroupVoiceCallSession(
                id = callId,
                hostUhid = sender.localUhid,
                state = GroupVoiceCallState.Active
            ).also { it.members.add(sender.localUhid) }
        }

        for (uhid in memberUhids) {
            session.members.add(uhid)
            val payload = encodeGroupSignaling(
                kind = "invite",
                callId = callId,
                fromUhid = sender.localUhid,
                toUhid = uhid,
                invitedUhids = memberUhids,
                keyGeneration = session.keyGenerationUInt
            )
            sender.send(groupSignalingPacket(uhid, payload), uhid)
        }
    }

    /**
     * Join the call [callId] (called by an invited member).
     * The host will receive the join notification and rotate the key.
     */
    suspend fun join(callId: UUID) {
        val session = sessions[callId] ?: return
        session.members.add(sender.localUhid)
        session.state = GroupVoiceCallState.Active

        val payload = encodeGroupSignaling(
            kind = "join",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = session.hostUhid
        )
        sender.send(groupSignalingPacket(session.hostUhid, payload), session.hostUhid)

        // If local node is host, rotate key immediately
        if (sender.localUhid == session.hostUhid) {
            rotateKey(session)
        }
    }

    /**
     * Leave the call [callId].
     */
    suspend fun leave(callId: UUID) {
        val session = sessions[callId] ?: return
        session.members.remove(sender.localUhid)
        val isHost = sender.localUhid == session.hostUhid

        // Notify remaining members
        val leavePayload = encodeGroupSignaling(
            kind = "leave",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = ""
        )
        broadcastToMembers(session, leavePayload)

        if (isHost) {
            // Host leaving: end the call
            endCall(callId)
        } else if (session.members.contains(session.hostUhid)) {
            // Host is still present; ask host to rotate key by sending leave directly
            // The host's handlePacket will trigger rotation
        }

        if (session.members.isEmpty() || isHost) {
            sessions.remove(callId)
            session.state = GroupVoiceCallState.Ended
            onMembershipChanged?.invoke(session)
        }
    }

    /**
     * (Host) Kick [targetUhid] from the call.
     */
    suspend fun kick(callId: UUID, targetUhid: String) {
        val session = sessions[callId] ?: return
        if (sender.localUhid != session.hostUhid) return
        session.members.remove(targetUhid)

        val payload = encodeGroupSignaling(
            kind = "kick",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = targetUhid,
            kickedUhid = targetUhid
        )
        // Notify the kicked user
        sender.send(groupSignalingPacket(targetUhid, payload), targetUhid)
        // Rotate key so kicked member cannot decrypt future frames
        rotateKey(session)
        onMembershipChanged?.invoke(session)
    }

    /**
     * Send an encoded audio frame to all call members.
     * [keyGeneration] must match the current session key generation.
     */
    suspend fun sendFrame(callId: UUID, encodedAudio: ByteArray, isSilence: Boolean, keyGeneration: UInt) {
        val session = sessions[callId]
        if (session == null || session.state != GroupVoiceCallState.Active) return

        val seq = frameSequence++.toUInt()
        val payload = encodeGroupVoiceFrame(
            callId = callId,
            sequence = seq,
            timestampMs = System.currentTimeMillis(),
            isSilence = isSilence,
            keyGeneration = keyGeneration,
            encodedPayload = encodedAudio
        )
        val packet = MeshPacket(
            type = PacketType.VoiceCall,
            sourceUhid = sender.localUhid,
            destinationUhid = "",
            ttl = AetherConstants.DEFAULT_TTL,
            priority = 64,
            payload = payload
        )
        // Send to each member except self
        for (member in session.members) {
            if (member != sender.localUhid) {
                sender.send(packet.copy(destinationUhid = member), member)
            }
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Inbound packet dispatcher
    // ─────────────────────────────────────────────────────────────────────

    suspend fun handlePacket(packet: MeshPacket) {
        when (packet.type) {
            PacketType.VoiceSignaling -> handleSignaling(packet)
            PacketType.VoiceCall -> handleFrame(packet)
            else -> {}
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Private helpers
    // ─────────────────────────────────────────────────────────────────────

    private suspend fun handleSignaling(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val kind = JsonReader.readString(json, "kind") ?: return
        val callIdStr = JsonReader.readString(json, "call_id") ?: return
        val callId = runCatching { UUID.fromString(callIdStr) }.getOrNull() ?: return
        val fromUhid = JsonReader.readString(json, "from_uhid") ?: return

        when (kind) {
            "invite" -> {
                val invitedUhids = JsonReader.readStringArray(json, "invited_uhids")
                val keyGen = (JsonReader.readLong(json, "key_generation") ?: 0L).toInt()
                val session = GroupVoiceCallSession(
                    id = callId,
                    hostUhid = fromUhid,
                    state = GroupVoiceCallState.Invited,
                    keyGeneration = keyGen
                )
                session.members.addAll(invitedUhids)
                session.members.add(fromUhid)
                sessions[callId] = session
                onInvited?.invoke(session)
            }
            "join" -> {
                val session = sessions[callId] ?: return
                session.members.add(fromUhid)
                // If we are the host, rotate key on member join
                if (sender.localUhid == session.hostUhid) {
                    rotateKey(session)
                }
                onMembershipChanged?.invoke(session)
            }
            "leave" -> {
                val session = sessions[callId] ?: return
                session.members.remove(fromUhid)
                if (sender.localUhid == session.hostUhid) {
                    rotateKey(session)
                }
                onMembershipChanged?.invoke(session)
            }
            "kick" -> {
                val session = sessions[callId] ?: return
                val kickedUhid = JsonReader.readString(json, "kicked_uhid")
                if (kickedUhid != null) session.members.remove(kickedUhid)
                if (sender.localUhid == kickedUhid) {
                    sessions.remove(callId)
                    session.state = GroupVoiceCallState.Ended
                }
                onMembershipChanged?.invoke(session)
            }
            "end" -> {
                val session = sessions.remove(callId) ?: return
                session.state = GroupVoiceCallState.Ended
                onMembershipChanged?.invoke(session)
            }
            "key_rotation" -> {
                val session = sessions[callId] ?: return
                val keyGen = (JsonReader.readLong(json, "key_generation") ?: 0L).toInt()
                session.keyGeneration = keyGen
                onKeyRotation?.invoke(callId, session.keyGenerationUInt)
            }
        }
    }

    private fun handleFrame(packet: MeshPacket) {
        val data = packet.payload
        // Minimum: 16 (CallId) + 4 (seq) + 8 (ts) + 1 (silence) + 4 (keyGen) = 33 bytes
        if (data.size < 33) return

        val uuidBuf = ByteBuffer.wrap(data, 0, 16).order(ByteOrder.BIG_ENDIAN)
        val callId = UUID(uuidBuf.long, uuidBuf.long)

        val buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        buf.position(16)
        @Suppress("UNUSED_VARIABLE") val seq = buf.int.toUInt()
        @Suppress("UNUSED_VARIABLE") val ts = buf.long
        val isSilence = buf.get() != 0.toByte()
        val keyGen = buf.int.toUInt()
        val encoded = if (buf.hasRemaining()) {
            val arr = ByteArray(buf.remaining())
            buf.get(arr)
            arr
        } else ByteArray(0)

        onFrameReceived?.invoke(callId, packet.sourceUhid, encoded, isSilence, keyGen)
    }

    private suspend fun rotateKey(session: GroupVoiceCallSession) {
        session.keyGeneration++
        val payload = encodeGroupSignaling(
            kind = "key_rotation",
            callId = session.id,
            fromUhid = sender.localUhid,
            toUhid = "",
            keyGeneration = session.keyGenerationUInt
        )
        broadcastToMembers(session, payload)
        onKeyRotation?.invoke(session.id, session.keyGenerationUInt)
    }

    private suspend fun endCall(callId: UUID) {
        val session = sessions.remove(callId) ?: return
        session.state = GroupVoiceCallState.Ended
        val payload = encodeGroupSignaling(
            kind = "end",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = ""
        )
        broadcastToMembers(session, payload)
        onMembershipChanged?.invoke(session)
    }

    private suspend fun broadcastToMembers(session: GroupVoiceCallSession, payload: ByteArray) {
        for (member in session.members) {
            if (member != sender.localUhid) {
                val pkt = groupSignalingPacket(member, payload)
                sender.send(pkt, member)
            }
        }
    }

    private fun groupSignalingPacket(toUhid: String, payload: ByteArray) = MeshPacket(
        type = PacketType.VoiceSignaling,
        sourceUhid = sender.localUhid,
        destinationUhid = toUhid,
        ttl = AetherConstants.DEFAULT_TTL,
        priority = 32,
        payload = payload
    )
}

// ─────────────────────────────────────────────────────────────────────────────
// Wire helpers
// ─────────────────────────────────────────────────────────────────────────────

internal fun encodeGroupVoiceFrame(
    callId: UUID,
    sequence: UInt,
    timestampMs: Long,
    isSilence: Boolean,
    keyGeneration: UInt,
    encodedPayload: ByteArray
): ByteArray {
    val buf = ByteBuffer.allocate(16 + 4 + 8 + 1 + 4 + encodedPayload.size)
    buf.order(ByteOrder.BIG_ENDIAN)
    buf.putLong(callId.mostSignificantBits)
    buf.putLong(callId.leastSignificantBits)
    buf.order(ByteOrder.LITTLE_ENDIAN)
    buf.putInt(sequence.toInt())
    buf.putLong(timestampMs)
    buf.put(if (isSilence) 1.toByte() else 0.toByte())
    buf.putInt(keyGeneration.toInt())
    buf.put(encodedPayload)
    return buf.array()
}

private fun encodeGroupSignaling(
    kind: String,
    callId: UUID,
    fromUhid: String,
    toUhid: String,
    invitedUhids: List<String>? = null,
    kickedUhid: String? = null,
    keyGeneration: UInt? = null
): ByteArray {
    val sb = StringBuilder()
    sb.append('{')
    sb.append("\"kind\":\"").append(jsonEscape(kind)).append("\",")
    sb.append("\"call_id\":\"").append(callId).append("\",")
    sb.append("\"from_uhid\":\"").append(jsonEscape(fromUhid)).append("\",")
    sb.append("\"to_uhid\":\"").append(jsonEscape(toUhid)).append('"')
    if (invitedUhids != null) {
        sb.append(",\"invited_uhids\":[")
        invitedUhids.forEachIndexed { i, u ->
            if (i > 0) sb.append(',')
            sb.append('"').append(jsonEscape(u)).append('"')
        }
        sb.append(']')
    }
    if (kickedUhid != null) sb.append(",\"kicked_uhid\":\"").append(jsonEscape(kickedUhid)).append('"')
    if (keyGeneration != null) sb.append(",\"key_generation\":").append(keyGeneration.toLong())
    sb.append('}')
    return sb.toString().toByteArray(Charsets.UTF_8)
}

File: typescript/src/voice/GroupVoiceCallService.ts
/**
 * Host-driven group voice call service.
 *
 * Only the host (creator) can invite, kick, or end the call.
 * All members exchange frames via unicast to every other member.
 * When membership changes the host increments keyGeneration and
 * broadcasts a key_rotation signaling message.
 *
 * Signaling uses PacketType.VoiceSignaling (JSON GroupVoiceSignalingMessage).
 * Frames use PacketType.VoiceCall (binary GroupVoiceFrame).
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_TTL, MAX_GROUP_VOICE_MEMBERS } from "../constants.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";
import { uuidToBytes, bytesToUuid } from "./VoiceCallService.js";

// ──────────────────────────────────────────────────────────────────
// Constants
// ──────────────────────────────────────────────────────────────────

const GROUP_VOICE_FRAME_PRIORITY = 64;
const GROUP_VOICE_SIGNALING_PRIORITY = 32;

// ──────────────────────────────────────────────────────────────────
// Wire types (snake_case)
// ──────────────────────────────────────────────────────────────────

type GroupVoiceSignalingKind =
  | "invite"
  | "join"
  | "leave"
  | "kick"
  | "end"
  | "key_rotation";

interface GroupVoiceSignalingMessage {
  kind: GroupVoiceSignalingKind;
  call_id: string;
  from_uhid: string;
  to_uhid: string;
  invited_uhids?: string[];
  kicked_uhid?: string;
  key_generation?: number;
}

// ──────────────────────────────────────────────────────────────────
// Public types
// ──────────────────────────────────────────────────────────────────

export type GroupCallState = "idle" | "active" | "ended";

export interface GroupCallInfo {
  callId: string;
  hostUhid: string;
  members: Set<string>;
  state: GroupCallState;
  keyGeneration: number;
  startedAt: Date;
}

export interface GroupFrameEvent {
  callId: string;
  fromUhid: string;
  sequence: number;
  timestampMs: number;
  isSilence: boolean;
  keyGeneration: number;
  encodedPayload: Uint8Array;
}

// ──────────────────────────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────────────────────────

export class GroupVoiceCallService {
  private readonly calls = new Map<string, GroupCallInfo>();

  // Event callbacks
  onGroupCallInvited?: (info: GroupCallInfo) => void;
  onGroupCallActive?: (info: GroupCallInfo) => void;
  onGroupCallEnded?: (info: GroupCallInfo) => void;
  onMembershipChanged?: (info: GroupCallInfo) => void;
  onGroupFrameReceived?: (event: GroupFrameEvent) => void;

  constructor(private readonly sender: IMeshSender) {}

  // ──────────────── host actions ───────────────────────────────────

  async invite(callId: string, memberUhids: string[]): Promise<void> {
    if (!callId) throw new Error("callId must not be empty");
    if (memberUhids.length === 0) throw new Error("memberUhids must not be empty");

    const local = this.sender.localUhid;
    const members = new Set<string>([local, ...memberUhids]);
    if (members.size > MAX_GROUP_VOICE_MEMBERS) {
      throw new Error(`Group call exceeds max members (${MAX_GROUP_VOICE_MEMBERS})`);
    }

    const info: GroupCallInfo = {
      callId,
      hostUhid: local,
      members,
      state: "active",
      keyGeneration: 0,
      startedAt: new Date(),
    };
    this.calls.set(callId, info);

    for (const uhid of memberUhids) {
      await this.sendSignaling(
        {
          kind: "invite",
          call_id: callId,
          from_uhid: local,
          to_uhid: uhid,
          invited_uhids: memberUhids,
          key_generation: info.keyGeneration,
        },
        uhid,
      );
    }

    this.onGroupCallActive?.(info);
  }

  async kick(callId: string, targetUhid: string): Promise<void> {
    const info = this.requireHost(callId);
    if (!info.members.has(targetUhid)) return;

    info.members.delete(targetUhid);
    await this.sendSignaling(
      {
        kind: "kick",
        call_id: callId,
        from_uhid: this.sender.localUhid,
        to_uhid: targetUhid,
        kicked_uhid: targetUhid,
      },
      targetUhid,
    );

    info.keyGeneration += 1;
    await this.broadcastKeyRotation(info);
    this.onMembershipChanged?.(info);
  }

  async endCall(callId: string): Promise<void> {
    const info = this.requireHost(callId);
    info.state = "ended";
    for (const uhid of info.members) {
      if (uhid === this.sender.localUhid) continue;
      await this.sendSignaling(
        { kind: "end", call_id: callId, from_uhid: this.sender.localUhid, to_uhid: uhid },
        uhid,
      );
    }
    this.onGroupCallEnded?.(info);
  }

  // ──────────────── member actions ─────────────────────────────────

  async join(callId: string): Promise<void> {
    const info = this.calls.get(callId);
    if (!info) return;
    const local = this.sender.localUhid;
    if (!info.members.has(local)) {
      info.members.add(local);
      this.onMembershipChanged?.(info);
    }
    await this.sendSignaling(
      { kind: "join", call_id: callId, from_uhid: local, to_uhid: info.hostUhid },
      info.hostUhid,
    );
  }

  async leave(callId: string): Promise<void> {
    const info = this.calls.get(callId);
    if (!info) return;
    const local = this.sender.localUhid;
    info.members.delete(local);
    info.state = "ended";
    for (const uhid of info.members) {
      await this.sendSignaling(
        { kind: "leave", call_id: callId, from_uhid: local, to_uhid: uhid },
        uhid,
      );
    }
    this.onGroupCallEnded?.(info);
  }

  // ──────────────── audio frame sending ───────────────────────────

  async sendFrame(
    callId: string,
    encodedAudio: Uint8Array,
    isSilence: boolean,
    keyGeneration: number,
  ): Promise<void> {
    const info = this.calls.get(callId);
    if (!info || info.state !== "active") return;

    const local = this.sender.localUhid;
    const payload = encodeGroupVoiceFrame(callId, encodedAudio, isSilence, keyGeneration);

    for (const uhid of info.members) {
      if (uhid === local) continue;
      const packet = new MeshPacket();
      packet.type = PacketType.VoiceCall;
      packet.sourceUhid = local;
      packet.destinationUhid = uhid;
      packet.ttl = DEFAULT_TTL;
      packet.priority = GROUP_VOICE_FRAME_PRIORITY;
      packet.payload = payload;
      await this.sender.send(packet, uhid);
    }
  }

  // ──────────────── inbound packet handling ───────────────────────

  async onPacket(packet: MeshPacket): Promise<void> {
    switch (packet.type) {
      case PacketType.VoiceSignaling:
        await this.handleSignaling(packet);
        break;
      case PacketType.VoiceCall:
        this.handleFrame(packet);
        break;
      default:
        break;
    }
  }

  // ──────────────── inspection ─────────────────────────────────────

  getCallInfo(callId: string): GroupCallInfo | undefined {
    return this.calls.get(callId);
  }

  getActiveCalls(): GroupCallInfo[] {
    return Array.from(this.calls.values()).filter((c) => c.state === "active");
  }

  // ──────────────── private helpers ─────────────────────────────────

  private async handleSignaling(packet: MeshPacket): Promise<void> {
    let msg: GroupVoiceSignalingMessage;
    try {
      msg = JSON.parse(
        new TextDecoder().decode(packet.payload),
      ) as GroupVoiceSignalingMessage;
    } catch {
      return;
    }
    if (!msg.call_id || !msg.kind) return;

    const { kind, call_id: callId, from_uhid: fromUhid } = msg;

    switch (kind) {
      case "invite": {
        if (this.calls.has(callId)) return;
        const allMembers = new Set<string>([fromUhid, ...(msg.invited_uhids ?? [])]);
        const info: GroupCallInfo = {
          callId,
          hostUhid: fromUhid,
          members: allMembers,
          state: "active",
          keyGeneration: msg.key_generation ?? 0,
          startedAt: new Date(),
        };
        this.calls.set(callId, info);
        this.onGroupCallInvited?.(info);
        break;
      }
      case "join": {
        const info = this.calls.get(callId);
        if (!info) break;
        if (!info.members.has(fromUhid)) {
          info.members.add(fromUhid);
          // Host rotates key on new member
          if (info.hostUhid === this.sender.localUhid) {
            info.keyGeneration += 1;
            await this.broadcastKeyRotation(info);
          }
          this.onMembershipChanged?.(info);
        }
        break;
      }
      case "leave":
      case "kick": {
        const info = this.calls.get(callId);
        if (!info) break;
        const removed = kind === "kick" ? (msg.kicked_uhid ?? fromUhid) : fromUhid;
        info.members.delete(removed);
        if (removed === this.sender.localUhid) {
          info.state = "ended";
          this.onGroupCallEnded?.(info);
        } else {
          this.onMembershipChanged?.(info);
        }
        break;
      }
      case "key_rotation": {
        const info = this.calls.get(callId);
        if (!info) break;
        if (msg.key_generation !== undefined) {
          info.keyGeneration = msg.key_generation;
        }
        break;
      }
      case "end": {
        const info = this.calls.get(callId);
        if (!info) break;
        info.state = "ended";
        this.onGroupCallEnded?.(info);
        break;
      }
      default:
        break;
    }
  }

  private handleFrame(packet: MeshPacket): void {
    const frame = decodeGroupVoiceFrame(packet.payload);
    if (!frame) return;
    const info = this.calls.get(frame.callId);
    if (!info || info.state !== "active") return;

    this.onGroupFrameReceived?.({
      callId: frame.callId,
      fromUhid: packet.sourceUhid,
      sequence: frame.sequence,
      timestampMs: frame.timestampMs,
      isSilence: frame.isSilence,
      keyGeneration: frame.keyGeneration,
      encodedPayload: frame.encodedPayload,
    });
  }

  private async sendSignaling(
    msg: GroupVoiceSignalingMessage,
    toUhid: string,
  ): Promise<void> {
    const body = new TextEncoder().encode(JSON.stringify(msg));
    const packet = new MeshPacket();
    packet.type = PacketType.VoiceSignaling;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = toUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = GROUP_VOICE_SIGNALING_PRIORITY;
    packet.payload = body;
    await this.sender.send(packet, toUhid);
  }

  private async broadcastKeyRotation(info: GroupCallInfo): Promise<void> {
    const local = this.sender.localUhid;
    for (const uhid of info.members) {
      if (uhid === local) continue;
      await this.sendSignaling(
        {
          kind: "key_rotation",
          call_id: info.callId,
          from_uhid: local,
          to_uhid: uhid,
          key_generation: info.keyGeneration,
        },
        uhid,
      );
    }
  }

  private requireHost(callId: string): GroupCallInfo {
    const info = this.calls.get(callId);
    if (!info) throw new Error(`Unknown call: ${callId}`);
    if (info.hostUhid !== this.sender.localUhid) {
      throw new Error("Only the host can perform this action");
    }
    return info;
  }
}

// ──────────────────────────────────────────────────────────────────
// Binary codec
// ──────────────────────────────────────────────────────────────────

/**
 * GroupVoiceFrame binary payload:
 * [16] CallId (UUID RFC4122 big-endian)
 * [4]  Sequence (uint32 little-endian)
 * [8]  TimestampMs (int64 little-endian)
 * [1]  IsSilence (0 or 1)
 * [4]  KeyGeneration (uint32 little-endian)
 * [N]  EncodedPayload
 */
let groupFrameSeq = 0;

export function encodeGroupVoiceFrame(
  callId: string,
  encodedPayload: Uint8Array,
  isSilence: boolean,
  keyGeneration: number,
): Uint8Array {
  const buf = new Uint8Array(16 + 4 + 8 + 1 + 4 + encodedPayload.length);
  const dv = new DataView(buf.buffer);

  uuidToBytes(callId, buf, 0);
  dv.setUint32(16, groupFrameSeq++ & 0xffffffff, true);
  dv.setBigInt64(20, BigInt(Date.now()), true);
  buf[28] = isSilence ? 1 : 0;
  dv.setUint32(29, keyGeneration >>> 0, true);
  buf.set(encodedPayload, 33);

  return buf;
}

export function decodeGroupVoiceFrame(data: Uint8Array): {
  callId: string;
  sequence: number;
  timestampMs: number;
  isSilence: boolean;
  keyGeneration: number;
  encodedPayload: Uint8Array;
} | null {
  if (data.length < 33) return null;
  const dv = new DataView(data.buffer, data.byteOffset, data.byteLength);

  const callId = bytesToUuid(data, 0);
  const sequence = dv.getUint32(16, true);
  const timestampMs = Number(dv.getBigInt64(20, true));
  const isSilence = data[28] !== 0;
  const keyGeneration = dv.getUint32(29, true);
  const encodedPayload = data.slice(33);

  return { callId, sequence, timestampMs, isSilence, keyGeneration, encodedPayload };
}
